source/opt/dead_branch_elim_pass.cpp

```diff
@@ -0,0 +1,271 @@
+  ir::Instruction* cInst = def_use_mgr_->GetDef(condId);
+  switch (cInst->opcode()) {
+    case SpvOpConstantFalse: {
+      *condVal = false;
+      *condIsConst = true;
+    } break;
+    case SpvOpConstantTrue: {
+      *condVal = true;
+      *condIsConst = true;
+    } break;
+    case SpvOpLogicalNot: {
+      bool negVal;
+      (void)GetConstCondition(cInst->GetSingleWordInOperand(0),
+          &negVal, condIsConst);
+      if (*condIsConst)
+        *condVal = !negVal;
+    } break;
+    default: {
+      *condIsConst = false;
+    } break;
+  }
+}
+
+void DeadBranchElimPass::AddBranch(uint32_t labelId, ir::BasicBlock* bp) {
+  std::unique_ptr<ir::Instruction> newBranch(
+    new ir::Instruction(SpvOpBranch, 0, 0,
+        {{spv_operand_type_t::SPV_OPERAND_TYPE_ID, {labelId}}}));
+  def_use_mgr_->AnalyzeInstDefUse(&*newBranch);
+  bp->AddInstruction(std::move(newBranch));
+}
+
+void DeadBranchElimPass::KillAllInsts(ir::BasicBlock* bp) {
+  bp->ForEachInst([this](ir::Instruction* ip) {
+    def_use_mgr_->KillInst(ip);
+  });
+}
+
+bool DeadBranchElimPass::GetConstConditionalBranch(ir::BasicBlock* bp,
+    ir::Instruction** branchInst, ir::Instruction** mergeInst,
+    bool *condVal) {
+  auto ii = bp->end();
+  --ii;
+  *branchInst = &*ii;
+  if ((*branchInst)->opcode() != SpvOpBranchConditional)
+    return false;
+  --ii;
+  *mergeInst = &*ii;
+  if ((*mergeInst)->opcode() != SpvOpSelectionMerge)
+    return false;
+  bool condIsConst;
+  (void) GetConstCondition(
+      (*branchInst)->GetSingleWordInOperand(kBranchCondConditionalIdInIdx),
+      condVal, &condIsConst);
+  return condIsConst;
+}
+
+bool DeadBranchElimPass::EliminateDeadBranches(ir::Function* func) {
+  // Traverse blocks in structured order
+  std::list<ir::BasicBlock*> structuredOrder;
+  ComputeStructuredOrder(func, &structuredOrder);
+  std::unordered_set<ir::BasicBlock*> elimBlocks;
+  bool modified = false;
+  for (auto bi = structuredOrder.begin(); bi != structuredOrder.end(); ++bi) {
+    // Skip blocks that are already in the elimination set
+    if (elimBlocks.find(*bi) != elimBlocks.end())
+      continue;
+    // Skip blocks that don't have constant conditional branch
+    ir::Instruction* br;
+    ir::Instruction* mergeInst;
+    bool condVal;
+    if (!GetConstConditionalBranch(*bi, &br, &mergeInst, &condVal))
+      continue;
+
+    // Replace conditional branch with unconditional branch
+    const uint32_t trueLabId =
+        br->GetSingleWordInOperand(kBranchCondTrueLabIdInIdx);
+    const uint32_t falseLabId =
+        br->GetSingleWordInOperand(kBranchCondFalseLabIdInIdx);
+    const uint32_t mergeLabId =
+        mergeInst->GetSingleWordInOperand(kSelectionMergeMergeBlockIdInIdx);
+    const uint32_t liveLabId = condVal == true ? trueLabId : falseLabId;
+    const uint32_t deadLabId = condVal == true ? falseLabId : trueLabId;
+    AddBranch(liveLabId, *bi);
+    def_use_mgr_->KillInst(br);
+    def_use_mgr_->KillInst(mergeInst);
```

Comment:
Well, not exactly.  What you can't do is break out of two Selection constructs.  You *can* break out of the middle of a Selection to the merge block of an enclosing loop.  So that's a branch that skips over a merge block in between out to an outer merge.
Example is generated from:
```
void main() {
  int x = 0;
  bool c, d;
  while (d) {
    if (c) {
      x = 1;
      break; // goes out to the y=x statement, skipping over the merge for this "if".
      x = 2;
    } else {
      x = 3;
      break; // goes out to the y = x statement.
      x = 4;
    }
  }
  int y = x; 
}
```

But this isn't the case you're handling.  So your conclusion is fine.